### ISSUE TITLE
Make `permitOnlyProjectScopedClusters` the default

### ISSUE THREAD

**Issue by crenshaw-dev** [argoproj/argo-cd#10549]
# Summary

@blakepettersson [added](https://github.com/argoproj/argo-cd/pull/10237) the `permitOnlyProjectScopedClusters` flag, because it allows the intuitive behavior (if there are only project-scoped clusters, those should be the only ones allowed). We should make that the default in 3.0.

# Motivation

The current behavior is counter-intuitive, which may cause security issues for admins who are unaware of the default behavior.

# Proposal

Default `permitOnlyProjectScopedClusters` to true, and document the change in 2.x -> 3.0 upgrade notes.

**Comment by cilindrox** [comment]
Is there any way to attach more than a single project to a cluster?

Looking at this from the shared cluster perspective, if `team-A` team owns `project-A` and is allowed to deploy only to `cluster-C`, how can we enable `team-B` in `project-B` to deploy to `cluster-C` with only `permitOnlyProjectScopedClusters`?

AFAICT a cluster currently can be bound to a single project or am I doing something wrong?

**Comment by crenshaw-dev** [comment]
@cilindrox correct, currently each project-scoped cluster may be bound to only one project.

**Comment by agaudreault** [comment]
What would be the behavior if the flag is removed and no clusters are scoped to the current project?

How to differentiate between "This project should not be allowed to use global clusters, but the current cluster configured is found" and "This is a project that needs to use one of the configured global clusters"?

From what I understand, "if there are only project-scoped clusters, those should be the only ones allowed" means that any project will be able to use global clusters.

**Comment by blakepettersson** [comment]
@agaudreault 

>What would be the behavior if the flag is removed and no clusters are scoped to the current project?

Any cluster can be used.

> How to differentiate between "This project should not be allowed to use global clusters, but the current cluster configured is found" and "This is a project that needs to use one of the configured global clusters"?

I'm not sure if I understand "but the current cluster configured is found", can you clarify? IIRC if `permitOnlyProjectScopedClusters` is used, only clusters that are a part of that very same project can be used. In other words, global clusters are not allowed.

> From what I understand, "if there are only project-scoped clusters, those should be the only ones allowed" means that any project will be able to use global clusters.

That's correct

**Comment by reggie-k** [comment]
I am testing to understand whether the same cluster can be referenced from different projects.

**Comment by cilindrox** [comment]
I think my issue with this is for projects that deploy to several clusters, eg: a global ApplicationSet (think external-dns, that needs to ship to every cluster managed by argo).

I'd put this AppSet on its own project, eg: `remote-clusters`, but given the 1:1 mapping, I cannot make the project ship only to these with `permitOnlyProjectScopedClusters`. Conversely, any cluster that is bound to the `remote-clusters` project cannot be used by other project, eg `myappteam` project cannot target the same clusters  scoped to the `remote-clusters` project.

 Did I get that right?

**Comment by blakepettersson** [comment]
> I'd put this AppSet on its own project, eg: `remote-clusters`, but given the 1:1 mapping, I cannot make the project ship only to these with permitOnlyProjectScopedClusters. 

Yes you can; `permitOnlyProjectScopedClusters` is only scoped within a given `AppProject`. If you have the `remote-clusters` project set to `permitOnlyProjectScopedClusters: false` then it can ship manifests to any cluster, project-scoped or otherwise.

> Conversely, any cluster that is bound to the remote-clusters project cannot be used by other project

There are no restrictions imposed on other projects from using clusters in `remote-clusters` if `remote-clusters` has `permitOnlyProjectScopedClusters: true`; the restrictions only apply within `remote-clusters`. If you set `permitOnlyProjectScopedClusters: true` on all projects, then indeed your restrictions would apply.

> eg `myappteam` project cannot target the same clusters scoped to the `remote-clusters` project.

That's true if the `myappteam` project has `permitOnlyProjectScopedClusters: true` set.

**Comment by cilindrox** [comment]
thanks a lot for the added context and details @blakepettersson

**Comment by nitishfy** [comment]
Can I take this up?

**Comment by blakepettersson** [comment]
One thing I brought up is that it might make more sense to have this as a kind of an enum, since if this flag is set to true _only_ project-scoped clusters can be used.

>If the flag were to be flipped around to being the default, I'd argue that it would make sense for it to do the same thing we do for repos, which is if no project-scoped cluster is available, we check if the given cluster destination is an unscoped cluster, if so then :+1:
>
>Perhaps `permitOnlyProjectScopedClusters` could instead be an enum: `strict` (=same behaviour as `permitOnlyProjectScopedClusters: true` today), `lax` (=same behaviour as `permitOnlyProjectScopedClusters: false` today), and `fallback` (lookup a project-scoped cluster with a given url /name, if that doesn't exist find an unscoped cluster with the same url)

It's likely flipping the default to `true` would cause too much breakage with the current implementation.

**Comment by reggie-k** [comment]
I have confirmed the following use cases and their action items with @blakepettersson, with respect to flipping `permitOnlyProjectScopedClusters` to `true`, those examples should also go in the docs:

team-a project: shared-cluster1 (global) => breaking change, need to set permitOnlyProjectScopedClusters: false on team-a project
team-b project: cluster-b (project-scoped) => works OOTB
team-c project: shared-cluster2 (global) => breaking change, need to set permitOnlyProjectScopedClusters: false on team-c project
team-d project: cluster-d (global but not used in any other project) => breaking change, need to set cluster-d as scoped to team-d project
team-e project: cluster-e (scoped), shared-cluster2 (global), cluster-ee (global but not used in any other project) => breaking change, need to set permitOnlyProjectScopedClusters: false on team-e project because of shared-cluster2. cluster-ee can remain global

**Comment by reggie-k** [comment]
@nitishfy It would be great if you could take this up!
@crenshaw-dev What do you think about @blakepettersson suggestion for an enum for `permitOnlyProjectScopedClusters`, instead of flipping the default? Fuller details about the breaking change and needed action items in my comment above.

**Comment by nitishfy** [comment]
waiting for @crenshaw-dev feedback before starting to work...

**Comment by reggie-k** [comment]
@blakepettersson what happens when a local cluster is shared among projects?
Does it behave the same like a remote cluster with respect to `permitOnlyProjectScopedClusters`?

**Comment by reggie-k** [comment]
Gathering user feedback on adoption of `permitOnlyProjectScopedClusters` and methodologies of sharing clusters between AppProjects, please vote **on this comment**
1. My managed clusters are mostly shared between AppProjects (most of the clusters have at least 2 AppProjects deploying to them). I haven't configured `permitOnlyProjectScopedClusters:true` at all or have configured it for only a small subset of my AppProjects. Vote 👎 
2. Most of my AppProjects deploy to separate clusters. I have configured `permitOnlyProjectScopedClusters:true` for most of my AppProjects. Vote 🚀 
3. Most of my AppProjects deploy to separate clusters. I haven't configured `permitOnlyProjectScopedClusters:true` at all or have configured it for only a small subset of my AppProjects. Vote 👍

**Comment by reggie-k** [comment]
After further discussion with @crenshaw-dev, making `permitOnlyProjectScopedClusters` the default requires more thought on implementation and documentation, so will be more relevant for 4.0.
